File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int row, col;
    void dfs(vector<vector<char>>& grid, int i, int j) {
        if (i < 0 || j < 0 || i >= row || j >= col || grid[i][j] == '0') {
            return;
        }
        grid[i][j] = '0';
        dfs(grid, i+1, j);
        dfs(grid, i, j+1);
        dfs(grid, i-1, j);
        dfs(grid, i, j-1);
    }
    int numIslands(vector<vector<char>>& grid) {
        row = grid.size();
        col = grid[0].size();
        int count = 0;
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (grid[i][j] == '1') {
                    count++;
                    dfs(grid, i, j);
                }
            }
        }
        return count;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def dfs(self, grid: List[List[str]], i : int, j : int, row : int, col : int):
        if i < 0 or j < 0 or i >= row or j >= col or grid[i][j] == '0':
            return
        grid[i][j] = '0'
        self.dfs(grid, i+1, j, row, col)
        self.dfs(grid, i-1, j, row, col)
        self.dfs(grid, i, j+1, row, col)
        self.dfs(grid, i, j-1, row, col)

    def numIslands(self, grid: List[List[str]]) -> int:
        row, col, count = len(grid), len(grid[0]), 0
        for i in range(row):
            for j in range(col):
                if grid[i][j] == '1':
                    self.dfs(grid, i, j, row, col)
                    count += 1
        return count